service/history/ndc/hsm_state_replicator.go

```diff
@@ -0,0 +1,276 @@
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+//go:generate mockgen -copyright_file ../../../LICENSE -package $GOPACKAGE -source $GOFILE -destination hsm_state_replicator_mock.go
+
+package ndc
+
+import (
+	"context"
+
+	commonpb "go.temporal.io/api/common/v1"
+	"go.temporal.io/api/serviceerror"
+	enumsspb "go.temporal.io/server/api/enums/v1"
+	historyspb "go.temporal.io/server/api/history/v1"
+	"go.temporal.io/server/common/log"
+	"go.temporal.io/server/common/log/tag"
+	"go.temporal.io/server/common/namespace"
+	"go.temporal.io/server/common/persistence"
+	"go.temporal.io/server/common/persistence/versionhistory"
+	serviceerrors "go.temporal.io/server/common/serviceerror"
+	"go.temporal.io/server/service/history/hsm"
+	"go.temporal.io/server/service/history/shard"
+	"go.temporal.io/server/service/history/workflow"
+	wcache "go.temporal.io/server/service/history/workflow/cache"
+)
+
+type (
+	HSMStateReplicator interface {
+		SyncHSMState(
+			ctx context.Context,
+			request *shard.SyncHSMRequest,
+		) error
+	}
+
+	HSMStateReplicatorImpl struct {
+		shardContext  shard.Context
+		workflowCache wcache.Cache
+		logger        log.Logger
+	}
+)
+
+func NewHSMStateReplicator(
+	shardContext shard.Context,
+	workflowCache wcache.Cache,
+	logger log.Logger,
+) *HSMStateReplicatorImpl {
+
+	return &HSMStateReplicatorImpl{
+		shardContext:  shardContext,
+		workflowCache: workflowCache,
+		logger:        log.With(logger, tag.ComponentHSMStateReplicator),
+	}
+}
+
+func (r *HSMStateReplicatorImpl) SyncHSMState(
+	ctx context.Context,
+	request *shard.SyncHSMRequest,
+) (retError error) {
+	namespaceID := namespace.ID(request.WorkflowKey.GetNamespaceID())
+	execution := &commonpb.WorkflowExecution{
+		WorkflowId: request.WorkflowKey.GetWorkflowID(),
+		RunId:      request.WorkflowKey.GetRunID(),
+	}
+
+	firstItem, err := versionhistory.GetFirstVersionHistoryItem(request.EventVersionHistory)
+	if err != nil {
+		return err
+	}
+	lastItem, err := versionhistory.GetLastVersionHistoryItem(request.EventVersionHistory)
+	if err != nil {
+		return err
+	}
+
+	workflowContext, release, err := r.workflowCache.GetOrCreateWorkflowExecution(
+		ctx,
+		r.shardContext,
+		namespaceID,
+		execution,
+		workflow.LockPriorityHigh,
+	)
+	if err != nil {
+		return err
+	}
+	defer func() { release(retError) }()
+
+	mutableState, err := workflowContext.LoadMutableState(ctx, r.shardContext)
+	if err != nil {
+		if _, isNotFound := err.(*serviceerror.NotFound); isNotFound {
+			return serviceerrors.NewRetryReplication(
+				"sync HSM state encountered workflow not found",
+				namespaceID.String(),
+				execution.GetWorkflowId(),
+				execution.GetRunId(),
+				firstItem.EventId,
+				firstItem.Version,
```

Comment:
I cannot remember this. You may just need last item event id and version

Reply:
Yeah you are right. The underlying resend range is exclusive-exclusive, so I can't use firstItem here. 
Will change to EmptyEventID and Version to resend from beginning. Will also +1 to the lastItemEventID here and two other places in the file.